### PR TITLE
TEST: Add xfailing test for [func-returns-value]

### PR DESCRIPTION
**Pull request by harahu**
Add an xfailing test that reproduces #14179.

### CODE REVIEW

**Comment by harahu** [comment]
Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?

**Comment by brianschubert** [comment]
> Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?

I don't know about a canonical way, but any comment that points to the issue number / URL would probably be fine.
There's some [existing test cases](https://github.com/search?q=repo%3Apython%2Fmypy++path%3Atest-data++%28%22see%22+OR+%22TODO%22+OR+%22issue%22%29&type=code) that do this in various ways.

**Comment by harahu** [comment]
> > Is there a canonical way to point back to the github issue from the test code, so that if the test passes at some point, it is easy to also close the issue?
> 
> I don't know about a canonical way, but any comment that points to the issue number / URL would probably be fine. There's some [existing test cases](https://github.com/search?q=repo%3Apython%2Fmypy++path%3Atest-data++%28%22see%22+OR+%22TODO%22+OR+%22issue%22%29&type=code) that do this in various ways.

I added the URL as a comment.

File: test-data/unit/check-errorcodes.test [hunk 1]
```diff
@@ -563,6 +563,15 @@ y = A().g()  # E: "g" of "A" does not return a value (it only ever returns None)
...
 
+[case testErrorCodeStaticmethodDoesNotReturnValue-xfail]
+# See: https://github.com/python/mypy/issues/14179
+class A:
+    @staticmethod
+    def g() -> None: pass
+
+y1 = A.g()  # E: "g" of "A" does not return a value (it only ever returns None)  [func-returns-value]
+y2 = A().g()  # E: "g" of "A" does not return a value (it only ever returns None)  [func-returns-value]
+
```
Comment (brianschubert) [inline test-data/unit/check-errorcodes.test:574]:
```suggestion
[builtins fixtures/staticmethod.pyi]

```
Missing fixture for `staticmethod`.  See details at https://github.com/python/mypy/blob/master/test-data/unit/README.md#fixtures

**Comment by brianschubert** [comment]
I think this test would be better suited for `check-expressions.test`. There's a block of similar cases with the prefix `testNoneReturn`.

`check-errorcodes.test` for tests related to error codes themselves.

**Comment by brianschubert** [comment]
This issue appear to affect any decorated function, so there may be value in including a test using an arbitrary dummy decorator, and with multiple levels of decoration.

For your consideration, please feel free to adapt as you see fit:
```python
# test-data/unit/check-expressions.test
[case testNoneReturnDecorated]
from typing import Any, Callable, TypeVar

F = TypeVar('F', bound=Callable[..., Any])

def deco(f: F) -> F:
    pass

@deco
@deco
def f() -> None: 
    pass

class A:
    @staticmethod
    def s() -> None:
        pass

if int():
    x = f()      # E: "f" does not return a value (it only ever returns None)
if int():
    x = A.s()    # E: "s" of "A" does not return a value (it only ever returns None)
if int():
    x = A().s()  # E: "s" of "A" does not return a value (it only ever returns None)
[builtins fixtures/staticmethod.pyi]
```

**Comment by harahu** [comment]
> I think this test would be better suited for `check-expressions.test`. There's a block of similar cases with the prefix `testNoneReturn`.
> 
> `check-errorcodes.test` is for tests related to error codes themselves.

Moved and renamed the test.

> This issue appear to affect any decorated function, so there may be value in including a test using an arbitrary dummy decorator, and with multiple levels of decoration.

Good catch! I expanded the test based on your example.

**Review by brianschubert — COMMENTED** [review]
Nice

The `__call__` case is interesting. There are some other tests that cover opaque callables in other kinds of expressions, like `f = A(); y = f()`. It seems that the relevant bit here is having the callee be a `CallExpr`

A few nits on the error messages:

**Review by brianschubert — APPROVED** [review]
LGTM

Both tests pass when I force all function calls to always return `None`.

**Comment by hauntsaninja** [comment]
Superseded by https://github.com/python/mypy/pull/18015

**Pull request closed without merging**